lib/src/useTernaryDarkMode/useTernaryDarkMode.ts

```diff
@@ -0,0 +1,65 @@
+import { useEffect, useState } from 'react'
+
+// See: https://usehooks-ts.com/react-hook/use-local-storage
+import { useLocalStorage } from '../useLocalStorage'
+// See: https://usehooks-ts.com/react-hook/use-media-query
+import { useMediaQuery } from '../useMediaQuery'
+
+import { useUpdateEffect } from '..'
+
+const COLOR_SCHEME_QUERY = '(prefers-color-scheme: dark)'
+
+enum TernaryDarkMode {
+  System = 'system',
+  Dark = 'dark',
+  Light = 'light',
+}
+interface UseTernaryDarkModeOutput {
+  isDarkMode: boolean
+  ternaryDarkMode: TernaryDarkMode
+  toggleTernaryDarkMode: () => void
+}
+
+function useTernaryDarkMode(): UseTernaryDarkModeOutput {
+  const isDarkOS = useMediaQuery(COLOR_SCHEME_QUERY)
+  const [ternaryDarkMode, setTernaryDarkMode] =
+    useLocalStorage<TernaryDarkMode>('ternaryDarkMode', TernaryDarkMode.System)
+  const [isDarkMode, setDarkMode] = useState<boolean>(isDarkOS)
+
+  // Update darkMode if os prefers changes
+  useUpdateEffect(() => {
+    if (ternaryDarkMode === TernaryDarkMode.System) {
+      setDarkMode(isDarkOS)
+    }
+  }, [isDarkOS])
+
+  useEffect(() => {
+    switch (ternaryDarkMode) {
+      case TernaryDarkMode.Light:
+        setDarkMode(false)
+        break
+      case TernaryDarkMode.System:
+        setDarkMode(isDarkOS)
+        break
+      case TernaryDarkMode.Dark:
+        setDarkMode(true)
+        break
+    }
+  }, [ternaryDarkMode, isDarkOS])
+
+  return {
+    isDarkMode,
+    ternaryDarkMode,
+    toggleTernaryDarkMode: () => {
+      setTernaryDarkMode((currMode: TernaryDarkMode) =>
+        currMode === TernaryDarkMode.Dark
+          ? TernaryDarkMode.Light
+          : currMode === TernaryDarkMode.Light
+          ? TernaryDarkMode.System
+          : TernaryDarkMode.Dark,
+      )
+    },
+  }
```

Comment:
- Please add a method like `setTernaryDarkMode` to allow more hook implementations.
- For `toggleTernaryDarkMode`, it's quite better to create a function in the hook body and pass it to the return statement. Why? For clarity and because this pattern could create memory leak because of the anonymous function that cannot be collected by the garbage collector.

```ts
const toggleTernaryDarkMode = () => {
  // -- snip --
}

const setTernaryDarkMode = () => {
  // -- snip --
}

return {
  // -- snip -- 
  toggleTernaryDarkMode,
  setTernaryDarkMode
}

Reply:
I didn't know about the memory leak. Maybe it would be a problem. Using a `map` to do it now.

Reply:
```TS
      const toggleMap = new Map<TernaryDarkMode, TernaryDarkMode>([
        ['light', 'system'],
        ['system', 'dark'],
        ['dark', 'light'],
      ])
      setTernaryDarkMode((tdm: TernaryDarkMode) => toggleMap.get(tdm))
```

I can't fix the problem with map. TS deduced it's a `TernaryDarkMode | undefined`. I'm going back with `object` and `Record<>`